### PR TITLE
Set conntrack params in kube-proxy

### LINKED ISSUE #18604: kube-proxy: increase conntrack limits, decrease timeouts
At least one user has had issues with this, and I expect more.  It manifests as dropped packets/connections and dmesg show:

```
[1454500.910215] nf_conntrack: table full, dropping packet
[1454500.911030] nf_conntrack: table full, dropping packet
[1454502.184942] nf_conntrack: table full, dropping packet
[1454502.186097] nf_conntrack: table full, dropping packet
[1454503.898933] nf_conntrack: table full, dropping packet
[1454503.899431] nf_conntrack: table full, dropping packet
[1454505.211093] nf_conntrack: table full, dropping packet
[1454505.211688] nf_conntrack: table full, dropping packet
[1454506.507374] nf_conntrack: table full, dropping packet
[1454506.539645] nf_conntrack: table full, dropping packet
[1454508.027134] nf_conntrack: table full, dropping packet
[1454508.027234] nf_conntrack: table full, dropping packet
[1454509.461080] nf_conntrack: table full, dropping packet
[1454509.461217] nf_conntrack: table full, dropping packet
```

Maybe something like:

```
X=$(cat /proc/sys/net/netfilter/nf_conntrack_max); echo $((X*4)) > /proc/sys/net/netfilter/nf_conntrack_max; echo -n "was $X; now "; cat /proc/sys/net/netfilter/nf_conntrack_max

X=$(cat /sys/module/nf_conntrack/parameters/hashsize); echo $((X*4)) > /sys/module/nf_conntrack/parameters/hashsize; echo -n "was $X; now "; cat /sys/module/nf_conntrack/parameters/hashsize

X=$(cat /proc/sys/net/netfilter/nf_conntrack_generic_timeout); echo $((X/4)) > /proc/sys/net/netfilter/nf_conntrack_generic_timeout; echo -n "was $X; now "; cat /proc/sys/net/netfilter/nf_conntrack_generic_timeout 

X=$(cat /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established); echo $((X/5)) > /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established; echo -n "was $X; now "; cat /proc/sys/net/netfilter/nf_conntrack_tcp_timeout_established
```

Or maybe we should monitor `nf_conntrack_count` and increase only when needed (it does consume some kernel memory)

Let's keep an eye on this for 1.2.

### PR DESCRIPTION
**Pull request by thockin**
Add flags to control max connections (set to 256k vs 64k default) and TCP
established timeout (set to 1 day vs 5 day default).  Flags can be set to 0 to
mean "don't change it".

This is only set at startup, and not wrapped in a rectifier loop.

Tested manually.

Fixes #18604

### CODE REVIEW

File: cmd/kube-proxy/app/conntrack.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	glog.Infof("Setting nf_conntrack_max to %d", max)
+	if err := sysctl.SetSysctl("net/netfilter/nf_conntrack_max", max); err != nil {
+		return err
+	}
+	// TODO: generify this and sysctl to a new sysfs.WriteInt()
+	glog.Infof("Setting conntrack hashsize to %d", max/4)
+	return ioutil.WriteFile("/sys/module/nf_conntrack/parameters/hashsize", []byte(strconv.Itoa(max/4)), 0640)
+}
+
+func (realConntracker) SetTCPEstablishedTimeout(seconds int) error {
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/conntrack.go]: should the type be Duration instead of int ?

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -323,6 +335,20 @@ func (s *ProxyServer) Run(_ []string) error {
 		}, 5*time.Second, util.NeverStop)
 	}
 
+	// Tune conntrack, if requested
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: not part of this change, but should s.birthCry(), line 326, be just before SyncLoop(), line 353 ?

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -51,24 +51,26 @@ import (
...
+	ProxyMode                      string
+	IptablesSyncPeriod             time.Duration
+	ConfigSyncPeriod               time.Duration
+	NodeRef                        *api.ObjectReference // Reference to this node.
+	MasqueradeAll                  bool
+	CleanupAndExit                 bool
+	KubeAPIQPS                     float32
+	KubeAPIBurst                   int
+	UDPIdleTimeout                 time.Duration
+	ConntrackMax                   int
```
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: should this be time.Duration

File: cmd/kube-proxy/app/server.go (hunk 2 above)
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/server.go]: looks good overall.. minor nit..

File: cmd/kube-proxy/app/conntrack.go (hunk 1 above)
Comment (thockin) [inline cmd/kube-proxy/app/conntrack.go]: I chose seconds because any finer granularity is not respected.  I could go either way, but given the very limited exposure of this, I think simpler is better.

File: cmd/kube-proxy/app/server.go (hunk 2 above)
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: done

File: cmd/kube-proxy/app/server.go (hunk 3 above)
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: As previous - it can't be any finer than seconds, so I thought simpler was better.  If you think it's clearer I can change it, but it will involve getting the value back out by .Seconds() (float64) and then casting to int.

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

File: cmd/kube-proxy/app/conntrack.go (hunk 1 above)
Comment (ArtfulCoder) [inline cmd/kube-proxy/app/conntrack.go]: sounds good

**Comment by ArtfulCoder** [comment]
@k8s-bot unit test this please

a unit test seems to have failed. re-running test to confirm

**Comment by thockin** [comment]
@k8s-bot unit test this please

A different test fails every time.  This is intolerable

**Comment by thockin** [comment]
w00t!  Green at last. Green at last!  Thank git alrighty it is green at last.'

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
This is hopeless - even if I get it to go green, the chances of it being green again for Jenkins is 0.

@k8s-bot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by karlkfi** [comment]
Kubernetes-Mesos CI smoke tests have been failing on master since this was merged. Guess that CI failure up there wasn't just a red herring! 

Haven't figured out why this broke the tests, but it didn't just break them, it stopped the JUnit test report from being produced too...

**Comment by karlkfi** [comment]
Mesos smoke tests are fixed by #19277. Kinda stupid that it doesn't work without the new flags specified.

**Comment by quinton-hoole** [comment]
@thockin @ArtfulCoder sorry to chime in so late on this one.  A few words of caution:

1. From my experience, increasing the size of the conntrack table does not often help matters. An application that initiates an essentially unlimited number connections in parallel (think web crawler) will typically blow this table, whether it's 64K, 256K or even larger. A better approach is to rate limit the outgoing connection establishment packets.
2. Increasing the conntrack table size requires significant extra kernel memory, so you typically need to tune some other kernel parameters too (I forget which) to make the larger table fit.  It might have fitted by pure luck in the particular kernel config that you tested on, but it's quite possible that it will cause other kernels to explode.  Distro managers (Redhat, Ubuntu etc) typically tune all the necessary parameters correctly to work in harmony, so messing with individual ones can be dangerous.

File: cmd/kube-proxy/app/conntrack.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+type realConntracker struct{}
+
+func (realConntracker) SetMax(max int) error {
+	glog.Infof("Setting nf_conntrack_max to %d", max)
+	if err := sysctl.SetSysctl("net/netfilter/nf_conntrack_max", max); err != nil {
+		return err
+	}
+	// TODO: generify this and sysctl to a new sysfs.WriteInt()
+	glog.Infof("Setting conntrack hashsize to %d", max/4)
+	return ioutil.WriteFile("/sys/module/nf_conntrack/parameters/hashsize", []byte(strconv.Itoa(max/4)), 0640)
```
Comment (jdef) [inline cmd/kube-proxy/app/conntrack.go]: this line appears to be the culprit that fails our smoke testing when `max` is non-zero.
xref https://github.com/mesosphere/kubernetes-mesos/issues/724

Comment (jdef) [inline cmd/kube-proxy/app/conntrack.go]: for posterity: the nf_conntrack module doesn't seem to support setting the value of this hashsize parameter for network namespace other than `init_net`; this is strictly incompatible with our mesos/docker-based testing environment.

**Comment by thockin** [comment]
On Tue, Jan 5, 2016 at 9:40 AM, Quinton Hoole <notifications@github.com> wrote:
>
> @thockin @ArtfulCoder sorry to chime in so late on this one. A few words of caution:
>
> From my experience, increasing the size of the conntrack table does not often help matters. An application that initiates an essentially unlimited number connections in parallel (think web crawler) will typically blow this table, whether it's 64K, 256K or even larger. A better approach is to rate limit the outgoing connection establishment packets.

I don't disagree.  UDP is particularly unpleasant because we just have
to leave it to linger.

> Increasing the conntrack table size requires significant extra kernel memory, so you typically need to tune some other kernel parameters too (I forget which) to make the larger table fit. It might have fitted by pure luck in the particular kernel config that you tested on, but it's quite possible that it will cause other kernels to explode. Distro managers (Redhat, Ubuntu etc) typically tune all the necessary parameters correctly to work in harmony, so messing with individual ones can be dangerous.

We tune the hash size, too, following best practices, though
apparently this breaks in Mesos.

File: cmd/kube-proxy/app/conntrack.go (hunk 4 above)
Comment (thockin) [inline cmd/kube-proxy/app/conntrack.go]: Thanks for tracking it - what is the fix?  What can we do?

Comment (jdef) [inline cmd/kube-proxy/app/conntrack.go]: for k8s-mesos I've disabled these tuning parameters by default (read: zero by default). this fixes our CI environment immediately. users can still tweak them if needed/wanted. short of changing the way hashsize is implemented in the kernel module i'm not sure how else to really "fix" this.

Comment (unicell) [inline cmd/kube-proxy/app/conntrack.go]: More links I collected for the context:

- k8s-mesos issue @jdef mentioned, that might be related to setting hashsize inside nested network namespace
  https://github.com/mesosphere/kubernetes-mesos/issues/724

- more general discussion regarding whether sysctl are namespace safe
  https://github.com/kubernetes/kubernetes/issues/29572

- initial Linux kernel change, that restrict conntrack hash resize to init_net
  https://lwn.net/Articles/375395/
